Fix Newtonian Euler characteristics to allow 1d EOS's

fmahebert opened the pull request:
## Proposed changes

The quantity `kappa_over_density` goes to zero for these equations of state. This commit changes the normalization of some of the eigenvectors that enter the characteristic transformation matrices to avoid dividing by this quantity.

The left and right eigenvector matrices only change to adjust the normalizations. The test changes more because I now check the matrices with both a Polytropic and an IdealFluid equation of state.

### Types of changes:

- [x] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Fix NewtEuler characteristics to allow 1d EoS

The quantity kappa_over_density goes to zero for these equations of
state. This commit changes the normalization of some eigenvectors to
avoid dividing by this quantity.

carmaza committed:
Have basis vector test helper handle doubles and DataVectors

moxcodes reviewed (dismissed):
Looks good

carmaza commented on src/Evolution/Systems/NewtonianEuler/Characteristics.cpp:
```diff
@@ -211,28 +211,26 @@ Matrix left_eigenvectors<1>(const tnsr::I<double, 1>& velocity,
-                       get(specific_enthalpy) +
-                       get(sound_speed_squared) / get(kappa_over_density);
-
+  const double u2 = get(dot_product(velocity, velocity));
```
`u` is one of the velocity components, so I'd just call this `velocity_squared` everywhere.

carmaza commented on tests/Unit/Evolution/Systems/NewtonianEuler/Test_Characteristics.cpp:
```diff
@@ -292,6 +268,61 @@ void test_right_and_left_eigenvectors() noexcept {
+  test_left_and_right_eigenvectors_impl(
+      density, velocity, specific_internal_energy, unit_normal, polytropic_eos);
+}
+
```
It might be worth checking along coordinate axes as well, for which you'd like to loop over directions, as done for the speeds.

fmahebert replied:
Good idea, I'll add a new commit to this PR that checks this

carmaza reviewed (commented):
Feel free to squash.

fmahebert committed:
Test NewtEuler characteristics aligned with axes

fmahebert commented:
@carmaza Thanks for taking a look! I've squashed in the variable rename and added a commit to test the characteristics aligned with the axes.

@moxcodes could you look again?

moxcodes commented on tests/Unit/Evolution/Systems/NewtonianEuler/Test_Characteristics.cpp:
```diff
@@ -318,6 +318,14 @@ void test_right_and_left_eigenvectors() noexcept {
       density, velocity, specific_internal_energy, unit_normal, ideal_gas_eos);
 
+  // Test also in case where unit normal is aligned to axes
+  for (size_t i = 0; i < Dim; ++i) {
```
Does the axis-aligned test also work for the polytrope?

fmahebert replied:
Yes. Pushed update.

moxcodes reviewed (dismissed):
looks good to me.

carmaza commented on tests/Unit/Evolution/Systems/NewtonianEuler/Test_Characteristics.cpp:
```diff
@@ -321,6 +321,18 @@ void test_right_and_left_eigenvectors() noexcept {
       density, velocity, specific_internal_energy, unit_normal, polytropic_eos);
+
+  // Test also in cases where unit normal is aligned to coordinate axes
+  for (size_t i = 0; i < Dim; ++i) {
```
```
for (const auto& direction : Direction<Dim>::all_directions()) {
  const auto aligned = euclidean_basis_vector(direction, used_for_size);
  test_left_and_right_eigenvectors_impl(
      density, velocity, specific_internal_energy, aligned, ideal_gas_eos);

  test_left_and_right_eigenvectors_impl(
      density, velocity, specific_internal_energy, aligned, polytropic_eos);
}
```

should give the test along all directions for all dimensions for the same number of lines. However, I noticed that `euclidean_basis_vector` only works for `DataVectors`, for which I pushed 4e473ff2b63775ebdfa605bb454403e015d9f661 in #1958 as I also use it for relativistic hydro. Feel free to cherry-pick that commit and update your PR, or leave it as it is.

fmahebert commented:
@carmaza I've cherry-picked your commit and updated the for loop as you suggested :)

fmahebert commented:
@sxs-collaboration/spectre-core-devs This should be ready for final review

nilsdeppe reviewed (approved):
![image](https://user-images.githubusercontent.com/13531030/74076431-03022780-49e6-11ea-9142-3bd5bbda3dab.png)
